### PR TITLE
Add series_merge.py

### PR DESCRIPTION
**Pull request by Arthasasyan**
Redmine Issue 6748
The aim is to develop a python script that can merge one series into another without creating new series. The script needs to determine first date for each destination series and copy source data only UNTIL such first date.

### CODE REVIEW

File: examples/series_merge.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    tag_expression = args.tag_expression[0]
+start_date = '1970-01-01T00:00:00Z'
+
+
+connection = connect('/path/to/connection.properties')
+
+entity_service = EntitiesService(connection)
+
+if entity_service.get(source_entity) is None:
+    raise NameError("'" + source_entity + "' entity does not exist")
```
Comment (annstriganova) [inline examples/series_merge.py]: I'm sure, there is no reason to throw an error, compare:
Selection_001

with

Selection_001

**Comment by rodionos** [comment]
What's the purpose of tag expression? If it's omitted, does the script copy all series under specified metric and src entity?

File: examples/series_merge.py (hunk 1 above)
Comment (Arthasasyan) [reply examples/series_merge.py]: New output for non-existent entity:

image

File: examples/series_merge.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+
+#Parse CLI args
+parser = argparse.ArgumentParser()
+parser.add_argument('--src_entity', nargs=1, required=True, help='Source entity')
+parser.add_argument('--dst_entity', nargs=1, required=True, help='Destination entity')
+parser.add_argument('--metric_name', nargs=1, required=True, help='Metric to be cloned')
+parser.add_argument('--tag_expression', nargs=1, required=False, help='Tag expression to be acquired for source series')
+args = parser.parse_args()
+source_entity = args.src_entity[0]
```
[the remark this answers is not in the document]
Comment (Arthasasyan) [reply examples/series_merge.py]: New help window:

```
$ python3.5 series_merge.py -h
usage: series_merge.py [-h] [--tag_expression TAG_EXPRESSION] --src_entity
                       SRC_ENTITY --dst_entity DST_ENTITY --metric_name
                       METRIC_NAME

Copy metric values for series with source entity to one withdestination entity

optional arguments:
  -h, --help            show this help message and exit
  --tag_expression TAG_EXPRESSION, -te TAG_EXPRESSION
                        Tag expression to match source series

required arguments:
  --src_entity SRC_ENTITY, -se SRC_ENTITY
                        Source entity
  --dst_entity DST_ENTITY, -de DST_ENTITY
                        Destination entity
  --metric_name METRIC_NAME, -mn METRIC_NAME
                        Metric to be cloned

```

**Comment by Arthasasyan** [comment]
@rodionos if `tag_expression` is omitted only series with tags set that is met in both source and destination series will be copied
The purpose of this option is to filter tags that should be searched in destination series and than copied from source

**Comment by Arthasasyan** [comment]
Log source series with full key: metric, entity, tags
Log first date for destination series and full key for destination series
Log warning if source series don't have matched destination series

**Comment by annstriganova** [comment]
For some reason I've got the following error:
Selection_001

Could it be some kind of circular dependency?

**Comment by Arthasasyan** [comment]
@annstriganova could not reproduce it

```
akost@svgntb100:~/atsd-api-python/examples$ python3.5 series_merge.py -se akost_test_source_ent -de akost_test_dst_ent -mn akost_test_metr -te "tags.tag_1='one'"
INFO:root:Destination series is 'akost_test_dst_ent' entity, 'akost_test_metr' metric, '{'tag_1': 'one'}' tags
INFO:root:First datetime for destination series: 2019-12-11T16:34:30+03:00
INFO:root:Source series is: 'akost_test_source_ent' entity, 'akost_test_metr' metric, '{'tag_1': 'one'}' tags
WARNING:root:No series found for 'akost_test_source_ent' entity, 'akost_test_metr' metric and '{'tag_1': 'one'}' before 2019-12-11T16:34:30+03:00
INFO:root:Destination series is 'akost_test_dst_ent' entity, 'akost_test_metr' metric, '{'tag_1': 'one', 'only_dst': 'dst'}' tags
INFO:root:First datetime for destination series: 2019-12-11T16:54:16.505+03:00
INFO:root:Source series is: 'akost_test_source_ent' entity, 'akost_test_metr' metric, '{'tag_1': 'one', 'only_dst': 'dst'}' tags
WARNING:root:No series found for 'akost_test_source_ent' entity, 'akost_test_metr' metric and '{'tag_1': 'one', 'only_dst': 'dst'}' before 2019-12-11T16:54:16.505+03:00
WARNING:root:No series found for 'akost_test_dst_ent' entity, 'akost_test_metr' metric and '{'tag_1': 'one', 'only_src': 'src'}'
```

Also could not find circular dependencies
Could it be problems with your installation?

**Comment by annstriganova** [comment]
Yes, problem was at my side.

**Comment by Arthasasyan** [comment]
Return `--start_time` option, it would be useful to ignore random dates

**Comment by Arthasasyan** [comment]
```
$ python3.5 series_merge.py -h
usage: series_merge.py [-h] [--tag_expression TAG_EXPRESSION] [--dry_run]
                       [--start_time START_TIME] --src_entity SRC_ENTITY
                       --dst_entity DST_ENTITY --metric_name METRIC_NAME

Copy metric values for series with source entity to one withdestination entity

optional arguments:
  -h, --help            show this help message and exit
  --tag_expression TAG_EXPRESSION, -te TAG_EXPRESSION
                        Tag expression to match source series
  --dry_run, -dr        Run script without sending series to ATSD
  --start_time START_TIME, -st START_TIME
                        Set start date for query

required arguments:
  --src_entity SRC_ENTITY, -se SRC_ENTITY
                        Source entity
  --dst_entity DST_ENTITY, -de DST_ENTITY
                        Destination entity
  --metric_name METRIC_NAME, -mn METRIC_NAME
                        Metric to be cloned
```

**Comment by Arthasasyan** [comment]
Add parameter for batching

**Comment by Arthasasyan** [comment]
@annstriganova could you take one ore look on this?

File: examples/series_merge.py [hunk 3]
```diff
@@ -111,8 +115,25 @@ def err(tags, time=None, entity=source_entity):
...
+    else:
+        size = len(target_series.data)
+        start_position = 0
+        itearaion = 1
+        while size > 0:
+            batch_len = min(size, batch_size)
+            batch_data = [target_series.data[i] for i in range(start_position, start_position + batch_len, 1)]
+            batch = Series(target_series.entity, target_series.metric, tags=target_series.tags,
+                           data=batch_data)
+            logging.info("Iteration %s: Sending %s series to ATSD" % (itearaion, batch_len))
```
Comment (annstriganova) [inline examples/series_merge.py]: Minor typo `itearaion` -> `iteration`

Comment (Arthasasyan) [reply examples/series_merge.py]: Fixed

File: examples/series_merge.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    target_series = get_series_with_tags(source_series, series.tags)
+
+    if target_series is None:
+        logging.warning(err(series.tags, first_dst_time))
+        continue
+
+    target_series.entity = dst_entity
+    if batch_size == 0:
+        if not dry_run:
+            series_service.insert(target_series)
```
Comment (annstriganova) [inline examples/series_merge.py]: Could we also notify user about `dry-run`? I mean additional `logging.warning` statement.

Comment (Arthasasyan) [reply examples/series_merge.py]: I added it to lines where insert should happen:


```python
            if not dry_run:
                series_service.insert(batch)
            else:
                logging.warning("Dry run enabled, series are not inserted.")
```

File: examples/series_merge.py [hunk 5]
```diff
@@ -118,6 +118,8 @@ def err(tags, time=None, entity=source_entity):
     if batch_size == 0:
         if not dry_run:
             series_service.insert(target_series)
+        else:
+            logging.warning("Dry run enabled, series are not inserted.")
```
Comment (annstriganova) [inline examples/series_merge.py]: Is it suitable to move it to separate function to remove duplicated code?

Comment (Arthasasyan) [reply examples/series_merge.py]: You're right, moved it

**Pull request merged**